### PR TITLE
Windows: Check for OS build number version Runtime

### PR DESCRIPTION
**Pull request by AskBojesen**
To fix https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/issues/852

I have also added an installer project to easily test on another computer.
Eg. I have tested the winconsole on a Windows 10 with build number = 19041

### CODE REVIEW

**Comment by janusw** [comment]
@CodeMusher You reported in #849 that there are `InvalidCastException` when targeting Windows builds > 22k, but running on builds < 22k. Can you confirm that this PR fixes those exceptions?

File: Source/Plugin.BLE/Windows/BleImplementation.cs [hunk 1]
```diff
@@ -20,6 +20,26 @@ public class BleImplementation : BleImplementationBase
...
+        {
+            get
+            {
+                if (runtimeBuildNumber is null)
+                {
+                    runtimeBuildNumber = Environment.OSVersion.Version.Build;
+                }
+                return runtimeBuildNumber.Value;
+            }
+        }
```
Comment (janusw) [inline Source/Plugin.BLE/Windows/BleImplementation.cs:41]: To be honest, I don't see much of a point in adding this property.
Logically, it is not related to BLE at all, but is simply a general property of the OS.
Practically, one could just query `Environment.OSVersion.Version.Build` directly, which is simpler (and IMHO preferable, unless there is a good reason for adding a new property in BleImplementation).

Comment (AskBojesen) [reply Source/Plugin.BLE/Windows/BleImplementation.cs:41]: The reason is that I were inspired from my own project, where I need a solution for .net framework.
beacuse Environment.OSVersion.Version.Build does not work on .Net framework
So I needed this special method below.,
I agree to simplify now we can use Environment.OSVersion.Version.Build
```
  public static int CurrentBuild()
  {
      var reg = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows NT\CurrentVersion");
      var currentBuildStr = (string)reg.GetValue("CurrentBuild");
      var currentBuild = int.Parse(currentBuildStr);
      return currentBuild;
  }
```

**Comment by CodeMusher** [comment]
Yes, it does!16 maj 2024 kl. 21:31 skrev Janus Weil ***@***.***>:﻿
@CodeMusher You reported in #849 that there are InvalidCastException when targeting Windows builds > 22k, but running on builds < 22k. Can you confirm that this PR fixes those exceptions?

File: Source/Plugin.BLE/Windows/BleImplementation.cs (hunk 1 above)
Comment (janusw) [reply Source/Plugin.BLE/Windows/BleImplementation.cs]: > The reason is that I were inspired from my own project, where I need a solution for .net framework. beacuse Environment.OSVersion.Version.Build does not work on .Net framework So I needed this special method below.

Thanks for the comment. Note: For Android version checks in Plugin.BLE we also use a slightly different mechanism on .NET 6 and above, see e.g. here:

https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/blob/c894e46d7b51c3ea4debb10320483b08eea17a32/Source/Plugin.BLE/Android/Adapter.cs#L211-L215

IIRC, then `Build.VERSION.SdkInt` should also work with the latest .NET versions, but `OperatingSystem.IsAndroidVersionAtLeast` is better at removing deprecation warnings in such conditional code.

> I agree to simplify now we can use Environment.OSVersion.Version.Build

Thanks for updating this. Looks much better to me now :)

**Review by janusw — APPROVED** [review]
So, yeah, this is ok with me (with the latest changes and the confirmation that it fixes the issue).

**Comment by janusw** [comment]
I'd tend toward releasing 3.1.0 now (today or tomorrow). I think the most urgent things are fixed and further bugfixes can still be added in subsequent 3.1.x releases.

@AskBojesen Is this ok with you, or do you want anything else to be included in 3.1.0? Do we need another rc?

**Comment by AskBojesen** [comment]
> I'd tend toward releasing 3.1.0 now (today or tomorrow). I think the most urgent things are fixed and further bugfixes can still be added in subsequent 3.1.x releases.
> 
> @AskBojesen Is this ok with you, or do you want anything else to be included in 3.1.0? Do we need another rc?

No thanks! I tried to reproduce disconnect hangs sometimes on Windows. But could not - did 200 connect/disconnect on two different devices.

Looking forward to the 3.1.0 release :-)

**Comment by CodeMusher** [comment]
Hi, would it be possible to get a fix for my issue regarding INDICATE characteristic on Windows into the release?Skickat från min iPhone19 maj 2024 kl. 11:23 skrev Ask Magnussen Bojesen ***@***.***>:﻿

I'd tend toward releasing 3.1.0 now (today or tomorrow). I think the most urgent things are fixed and further bugfixes can still be added in subsequent 3.1.x releases.
@AskBojesen Is this ok with you, or do you want anything else to be included in 3.1.0? Do we need another rc?

No thanks! I tried to reproduce disconnect hangs sometimes on Windows. But could not - did 200 connect/disconnect on two different devices.
Looking forward to the 3.1.0 release :-)

**Comment by CodeMusher** [comment]
Yes, #847 is the one. I’ve written a solution in the issue that works for my use case. @AskBojesen do you think that is acceptable or could a characteristic have both INDICATE and NOTIFY?

**Comment by janusw** [comment]
> @AskBojesen do you think that is acceptable or could a characteristic have both INDICATE and NOTIFY?

Please continue this discussion in #847 (or the PR you'll open to fix it). It is highly off-topic here ;)

**Pull request merged**